File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/ActivePlayer.java
package com.lasthopesoftware.bluewater.client.playback.engine;


import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayingFile;
import com.lasthopesoftware.bluewater.client.playback.playlist.IPlaylistPlayer;
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager;
import com.namehillsoftware.handoff.promises.Promise;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observables.ConnectableObservable;

public class ActivePlayer implements IActivePlayer, AutoCloseable {

	private final Disposable fileChangedObservableConnection;
	private final IPlaylistPlayer playlistPlayer;
	private final ConnectableObservable<PositionedPlayingFile> observableProxy;

	public ActivePlayer(IPlaylistPlayer playlistPlayer, PlaylistVolumeManager volumeManagement) {
		this.playlistPlayer = playlistPlayer;

		volumeManagement.managePlayer(playlistPlayer);

		observableProxy = Observable.create(playlistPlayer).replay(1);

		fileChangedObservableConnection = observableProxy.connect();
	}

	@Override
	public Promise<?> pause() {
		return playlistPlayer.pause();
	}

	@Override
	public Promise<?> resume() {
		return playlistPlayer.resume();
	}

	@Override
	public ConnectableObservable<PositionedPlayingFile> observe() {
		return observableProxy;
	}

	@Override
	public void close() {
		fileChangedObservableConnection.dispose();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/ActivePlayer.kt
package com.lasthopesoftware.bluewater.client.playback.engine

import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayingFile
import com.lasthopesoftware.bluewater.client.playback.playlist.IPlaylistPlayer
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager
import com.namehillsoftware.handoff.promises.Promise
import io.reactivex.rxjava3.core.Observable

class ActivePlayer(
    private val playlistPlayer: IPlaylistPlayer,
    volumeManagement: PlaylistVolumeManager
) : IActivePlayer {
	private val observableProxy = Observable.create(playlistPlayer).replay(1).refCount()

    init {
        volumeManagement.managePlayer(playlistPlayer)
    }

    override fun pause(): Promise<*> = playlistPlayer.pause()

    override fun resume(): Promise<PositionedPlayingFile?> = playlistPlayer.resume()

    override fun observe(): Observable<PositionedPlayingFile> = observableProxy

	override fun halt(): Promise<*> = playlistPlayer.haltPlayback()
}
